Common++/header/Logger.h

```diff
@@ -284,17 +412,160 @@ namespace pcpp
+		 * @param source The log source.
+		 * @return A new LogContext.
+		 */
+		std::unique_ptr<internal::LogContext> createLogContext(LogLevel level, LogSource const& source = {});
+
+		/**
+		 * @brief Directly emits a log message bypassing all level checks.
+		 * @param source The log source.
+		 * @param level The log level for this message. This is only used for the log printer.
+		 * @param message The log message.
+		 */
+		void emit(LogSource const& source, LogLevel level, std::string const& message);
+		/**
+		 * @brief Directly emits a log message bypassing all level checks.
+		 * @param message The log message.
+		 */
+		void emit(std::unique_ptr<internal::LogContext> message);
+
+		/**
+		 * @brief Logs a message with the given source, level, and message.
+		 * @param message The log message.
+		 */
+		void log(std::unique_ptr<internal::LogContext> message);
+
+		/**
+		 * @brief Logs an object with the given source, level.
+		 *
+		 * The object is converted to a string via the std::ostream << operator.
+		 * @tparam T The type of object to be logged.
+		 * @param source The log source.
+		 * @param level The log level for this message.
+		 * @param message The object to be logged.
+		 */
+		template <class T> void log(LogSource const& source, LogLevel level, T const& message)
+		{
+			if (shouldLog(level, source.logModule))
+			{
+				auto ctx = createLogContext();
+				(*ctx) << message;
+				emit(std::move(ctx));
+			}
+		};
+
+		/**
+		 * @brief Logs an object with the given source at the Error level.
+		 *
+		 * The object is converted to a string via the std::ostream << operator.
+		 * @tparam T The type of object to be logged.
+		 * @param source The log source.
+		 * @param message The object to be logged.
+		 */
+		template <class T> void logError(LogSource const& source, T const& message)
+		{
+			log(source, LogLevel::Error, message);
+		};
+
+		/**
+		 * @brief Logs an object with the given source at the Info level.
+		 *
+		 * The object is converted to a string via the std::ostream << operator.
+		 * @tparam T The type of object to be logged.
+		 * @param source The log source.
+		 * @param message The object to be logged.
+		 */
+		template <class T> void logInfo(LogSource const& source, T const& message)
+		{
+			log(source, LogLevel::Info, message);
+		};
+
+		/**
+		 * @brief Logs an object with the given source at the Debug level.
+		 *
+		 * The object is converted to a string via the std::ostream << operator.
+		 * @tparam T The type of object to be logged.
+		 * @param source The log source.
+		 * @param message The object to be logged.
+		 */
+		template <class T> void logDebug(LogSource const& source, T const& message)
+		{
+			log(source, LogLevel::Debug, message);
+		};
+
 	private:
 		bool m_LogsEnabled;
-		Logger::LogLevel m_LogModulesArray[NumOfLogModules];
+		LogLevel m_LogModulesArray[NumOfLogModules];
 		LogPrinter m_LogPrinter;
 		std::string m_LastError;
```

Comment:
@seladb Do we absolutely need to store the last error? Because the way it works, if the logger is made to be used by more than one thread the value is completely useless as it can be changed at any time by any thread.